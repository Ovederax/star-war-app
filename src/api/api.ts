import axiosInstance, { AxiosResponse } from './axios'

export interface People {
  name: string
  height?: string
  mass?: string
  hair_color?: string
  skin_color?: string
  eye_color?: string
  birth_year?: string
  gender?: string
  homeworld: string
  films?: string[]
  species?: string[]
  vehicles?: string[]
  starships?: string[]
  created: string
  edited: string
  url: string
}

export interface PeoplesResult {
  count: number
  next: string
  previous: string | null
  results: People[]
}

interface GetAllParams {
  page?: number
  format?: 'wookiee'
}

type GetAllResponse = AxiosResponse<PeoplesResult>

export const getPeoplesData = async (params: GetAllParams): Promise<GetAllResponse> => {
  return axiosInstance.get<GetAllParams, GetAllResponse>('/people/', {
    params,
  })
}

interface GetParams {
  id?: number
}

type GetResponse = AxiosResponse<People>

export const getPeopleData = async (params: GetParams): Promise<GetResponse> => {
  return axiosInstance.get<GetParams, GetResponse>('/people/', {
    params,
  })
}
