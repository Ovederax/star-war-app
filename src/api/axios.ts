import axios from 'axios'

export type AxiosResponse<T> = {
  data: T
}

const backendURL = 'https://swapi.dev/api/'


const defaultOptions = {
  baseURL: backendURL,
}

const axiosInstance = axios.create(defaultOptions)

export default axiosInstance