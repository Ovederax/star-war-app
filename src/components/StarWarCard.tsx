import React from 'react'
import { Card, Stack, Typography } from '@mui/material'
import { People } from '../api/api'
import { getGenderBg, isExist } from '../utils'
import { Tag } from './Tag'
import Parameter from './Parameter'

interface Props {
  data: People
  showModal: (people: People) => void
}

const fontFamilyKarla = `'Karla', 'Roboto'`

export const testIdGender = 'card-gender-tag'
export const testIdBirthYear = 'card-birth-year-tag'

const StarWarCard = (props: Props) => {
  const { data, showModal } = props
  const { name, gender, birth_year, height, mass } = data

  const onShowModal = () => {
    showModal(data)
  }

  return (
    <Card
      sx={{ minWidth: 240, minHeight: '100%', cursor: 'pointer', background: '#F0F0F0' }}
      elevation={3}
      onClick={onShowModal}
    >
      <Stack spacing={1.5} pt={1} pb={1.5} px={3}>
        <Typography fontSize="18px" color="#212121" fontWeight="bold" fontFamily={fontFamilyKarla}>
          {name}
        </Typography>

        {(isExist(height) || isExist(mass)) && (
          <Stack direction="row" spacing={1.5}>
            {isExist(height) && <Parameter value={height} name="height" />}
            {isExist(mass) && <Parameter value={mass} name="mass" />}
          </Stack>
        )}

        {(isExist(gender) || isExist(birth_year)) && (
          <Stack direction="row" spacing={1.5}>
            {isExist(gender) && (
              <Tag data-testid={testIdGender} style={{ background: getGenderBg(gender) }}>
                {gender}
              </Tag>
            )}
            {isExist(birth_year) && (
              <Tag data-testid={testIdBirthYear} style={{ background: '#07D6F2' }}>
                {birth_year}
              </Tag>
            )}
          </Stack>
        )}
      </Stack>
    </Card>
  )
}

export default StarWarCard
