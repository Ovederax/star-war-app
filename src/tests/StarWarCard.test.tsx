import React from 'react'
import { render, screen } from '@testing-library/react'
import StarWarCard, { testIdBirthYear, testIdGender } from '../components/StarWarCard'
import { People } from '../api/api'

export const peopleWithTextContent: People = {
  name: 'Luke Skywalker',
  height: '172',
  mass: '77',
  hair_color: 'blond',
  skin_color: 'fair',
  eye_color: 'blue',
  birth_year: '19BBY',
  gender: 'male',
  homeworld: 'https://swapi.dev/api/planets/1/',
  created: '2014-12-09T13:50:51.644000Z',
  edited: '2014-12-20T21:17:56.891000Z',
  url: 'https://swapi.dev/api/people/1/',
}

test('SHOULD render parameters', () => {
  const fn = jest.fn()
  const { asFragment } = render(<StarWarCard data={peopleWithTextContent} showModal={fn} />)

  const elements = [
    screen.getByText(peopleWithTextContent.height || ''),
    screen.getByText(peopleWithTextContent.mass || ''),
    screen.getByText(peopleWithTextContent.gender || ''),
    screen.getByText(peopleWithTextContent.birth_year || ''),
    screen.getByTestId(testIdGender),
    screen.getByTestId(testIdBirthYear),
  ]

  for (const it of elements) {
    expect(it).toBeInTheDocument()
  }
  expect(asFragment()).toMatchSnapshot()
})

export const peopleWithoutTextContent: People = {
  name: 'Luke Skywalker',
  height: 'n/a',
  mass: 'unknown',
  gender: 'n/a',
  homeworld: 'https://swapi.dev/api/planets/1/',
  created: '2014-12-09T13:50:51.644000Z',
  edited: '2014-12-20T21:17:56.891000Z',
  url: 'https://swapi.dev/api/people/1/',
}

test('SHOULD NOT render parameters WHEN have not data', () => {
  const fn = jest.fn()
  const { asFragment } = render(<StarWarCard data={peopleWithoutTextContent} showModal={fn} />)

  const elements = [
    screen.queryByText('height'),
    screen.queryByText('mass'),
    screen.queryByTestId(testIdGender),
    screen.queryByTestId(testIdBirthYear),
  ]

  for (const it of elements) {
    expect(it).not.toBeInTheDocument()
  }
  expect(asFragment()).toMatchSnapshot()
})

test('SHOULD invoke showModal WHEN user click on card', () => {
  const fn = jest.fn()
  render(<StarWarCard data={peopleWithoutTextContent} showModal={fn} />)
  screen.getByText(peopleWithoutTextContent.name).click()
  expect(fn).toBeCalledTimes(1)
})
