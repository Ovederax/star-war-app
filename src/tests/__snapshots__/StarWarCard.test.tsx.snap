// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SHOULD NOT render parameters WHEN have not data 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation3 MuiCard-root css-17hoyqt-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiStack-root css-gqf4f0-MuiStack-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-npuat8-MuiTypography-root"
      >
        Luke Skywalker
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SHOULD render parameters 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation3 MuiCard-root css-17hoyqt-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiStack-root css-gqf4f0-MuiStack-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-npuat8-MuiTypography-root"
      >
        Luke Skywalker
      </p>
      <div
        class="MuiStack-root css-1r6rngd-MuiStack-root"
      >
        <div
          class="MuiStack-root css-zgbp5c-MuiStack-root"
        >
          <div
            class="css-18wfr7y"
          >
            172
          </div>
          <p
            class="MuiTypography-root MuiTypography-body1 css-794qm8-MuiTypography-root"
          >
            height
          </p>
        </div>
        <div
          class="MuiStack-root css-zgbp5c-MuiStack-root"
        >
          <div
            class="css-18wfr7y"
          >
            77
          </div>
          <p
            class="MuiTypography-root MuiTypography-body1 css-794qm8-MuiTypography-root"
          >
            mass
          </p>
        </div>
      </div>
      <div
        class="MuiStack-root css-1r6rngd-MuiStack-root"
      >
        <div
          class="css-1vheds3"
          data-testid="card-gender-tag"
          style="background: rgb(115, 214, 119);"
        >
          male
        </div>
        <div
          class="css-1vheds3"
          data-testid="card-birth-year-tag"
          style="background: rgb(7, 214, 242);"
        >
          19BBY
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
